hooks/environment

```diff
@@ -47,8 +60,35 @@ function version_a_gte_b() {
   return 0                              # major same, minor same, patch same or more
 }
 
-# Retries a command on failure.
+# Bash's -E option propagates error traps into called functions. In the case of the `_retry` function, we don't want to do that because if we interrupt execution on an error, it becomes impossible to retry.
+#
+# We need to do this in a wrapper because by the time we are executing _retry, it's too late to unset the error trap.
+#
 function retry() {
+  local -r -i retries="$1"; shift
+  local exit_code
+  local stdin_value
+  local with_stdin_flag=""
+
+  if [[ "$1" == "--with-stdin" ]]; then
+    with_stdin_flag=" --with-stdin"
+    read -sr stdin_value
+    shift
+  fi
+
+  [[ "$IS_BIG_E_OPTION_SET" == "yes" ]] && set +E
+
+  #shellcheck disable=SC2086
+  _retry "$retries" ${with_stdin_flag} "$@" <<< "${stdin_value:-}"
```

Comment:
A way to avoid disabling a very valid check would be to do something like the following:

```suggestion
  cmd=(_retry "$retries")
  if [[ "$1" == "--with-stdin" ]]; then
    cmd=+(--with-stdin)
    read -sr stdin_value
    shift
  fi

  [[ "$IS_BIG_E_OPTION_SET" == "yes" ]] && set +E

  "${cmd[@]}" "$@" <<< "${stdin_value:-}"
```